lib/pos_block_producer.go

```diff
@@ -75,7 +83,7 @@ func (pbp *PosBlockProducer) CreateUnsignedTimeoutBlock(latestBlockView *UtxoVie
 func (pbp *PosBlockProducer) createBlockTemplate(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
 	proposerRandomSeedSignature *bls.Signature) (BlockTemplate, error) {
 	// First get the block without the header.
-	currentTimestamp := time.Now().UnixNano()
+	currentTimestamp := pbp.timeSource.AdjustedTime().UnixNano()
```

Comment:
We shouldn't rely on a timeSource anymore. For context, Bitcoin has a really weird way of dealing with time that we sortof emulated before, but that we're not going to be porting over to PoS.

For PoS, we want to simply set the block timestamp to:
`currentTimestamp = MIN(time.UnixNano(), parentBlock.TstampNanos + 1)`

Critically, the nodes on the network should allow for some amount of drift, say a few minutes, but should reject blocks outside of that strict time range. Bitcoin does this rejection but its window is 2 hours or so, and it does not enforce monotonically increasing timestamps, which causes complications (it would be better if it did).

I remember we decided that monotonically increasing the timestamp would prevent attacks, but now that I'm thinking about it I'm not 100% sure we need it. It prevents attacks in Proof of Work because the timestamp is used by Proof of Work to adjust the difficulty, but I don't think we rely explicitly on the timestamp for anything in consensus itself, and so I think just having it be `currentTimestamp = time.UnixNano()` would probably work. In this case, a validator can fudge the time up to our strict time window for one block, but then the next validator would correct it.

The above being said, I recommend keeping monotonically increasing timestamps unless we have a good reason not to have them. Among other things, they lead to more consistent behavior for things that rely on the timestamp, such as lockups. In the specific example of lockups, you could have a weird case where a lockup is unlockable for one block but not unlockable for the next, which is kinda dumb (though not fatal).

So I would just leave a comment explaining as much as you can about the relevant context above.

Also just as a history lesson, I believe Bitcoin Core sets its timestamp based on the median of what its peers tell it the time is or something like that. I don't know if it still does this, but I think that's how Satoshi had it. It prevents your clock from drifting too hard from everyone else's clock, so that you're always suggesting blocks that are within their window.

Reply:
We settled on having strictly increasing block timestamps with up to a 10 minute forward buffer. It should not be possible for a block proposer to roll back time with a previously proposed block. This would break the integrity of lockups, among what I'm sure will be other edge cases built on future assumptions of strictly increasing timestamps in other use cases.

I'll update this to get rid of the timeSource, and to introduce strictly increasing timestamps.